### PR TITLE
fixes for psh kill working

### PR DESCRIPTION
**Pull request by damianloew**
## Description

-syscalls: fixes to prevent exceptions in syscalls_signalPost
-hal: ia32: apply clang-format to cpu.c file
-hal: ia32: fix calling hal_cpuReschedule with NULL arguments

## Motivation and Context

Killing processes in using psh command returned exceptions. These changes solve problems with kill.
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: (https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/55, https://github.com/phoenix-rtos/phoenix-rtos-project/pull/196).
- [x] Tested by hand on: (ia32-generic - qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/ia32/cpu.c [hunk 1]
```diff
@@ -135,36 +135,16 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
...
-		"1:;"
 		"pushf;"
-
-		"2:;"
 		"pushl %%cs;"
 		"cli;"
-		"leal 3f, %%eax;"
+		"leal 1f, %%eax;"
 		"pushl %%eax;"
 		"movl $0, %%eax;"
```
Comment (kemonats) [inline hal/ia32/cpu.c:149]:
```suggestion
		"xorl %%eax, %%eax;"
```
Faster and less bytes.

File: hal/ia32/cpu.c [hunk 2]
```diff
@@ -175,21 +155,21 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
 		"pushl $0;"
 		"pushl %%eax;"
 		"pushl $0;"
-		"movl %4, %%eax;"
+		"movl %1, %%eax;"
 		"call *%%eax;"
 		"cli;"
 		"addl $12, %%esp;"
 		"popl %%esp;"
```
Comment (kemonats) [inline hal/ia32/cpu.c:162]:
```suggestion
		"popl %%esp;"
```
Adding `12` to the `esp` and then loading the value from the stack to `esp`?

File: hal/ia32/cpu.c [hunk 3]
```diff
@@ -175,21 +155,21 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
...
 		"call _interrupts_multilockClear;"
- 		"pushl %%esp;"
+		"pushl %%esp;"
 
 		"jmp interrupts_popContext;"
 
-		"3:;"
+		"1:;"
 		"movl %%eax, %0"
 	: "=g" (err)
```
Comment (kemonats) [inline hal/ia32/cpu.c:171]:
```suggestion
	: "=a" (err)
```
You avoid contruction type:
```
mov %eax, %ecx
mov %ecx, %eax
```
If you accept this suggestion remove `eax` from `clobbers`.

File: hal/ia32/cpu.c [hunk 4]
```diff
@@ -175,21 +155,21 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
...
-	: "m" (spinlock), "m" (spinlock->lock), "r" (*scp), "g" (threads_schedule)
+	: "g" (threads_schedule)
 	: "eax", "edx", "esp", "cc", "memory");
```
Comment (kemonats) [inline hal/ia32/cpu.c:173]:
```suggestion
	: "eax", "esp", "cc", "memory");
```
`edx` is not used in the `asm` block.

**Review by kemonats — COMMENTED** [review]
My suggestions are beyond this PR, but since you are changing the code in this function I give them.

File: hal/ia32/cpu.c (hunk 1 above)
Comment (Maxez) [reply hal/ia32/cpu.c]: I agree, added in next commit.

File: hal/ia32/cpu.c (hunk 2 above)
Comment (Maxez) [reply hal/ia32/cpu.c]: We're adding 12 to `popl` 3 arguments `pushl`ed for `threads_schedule()` call. Doing `popl %%esp` and `pushl %%esp` (after `call _interrupts_multilockClear`) seems unnecessary so I've removed it (btw. it didn't hurt as `esp` is properly saved earlier with `call interrupts_pushContext`).

File: hal/ia32/cpu.c (hunk 3 above)
Comment (Maxez) [reply hal/ia32/cpu.c]: I agree, I also removed last line `movl %%eax, %0` as it's no longer needed.

File: hal/ia32/cpu.c (hunk 4 above)
Comment (Maxez) [reply hal/ia32/cpu.c]: It's there due to cdecl calling convention used for `threads_schedule()` - `eax`, `ecx` and `edx` aren't saved by the callee-side and could be modified by `threads_schedule()`.
Added missing `ecx` in clobber list.

File: hal/ia32/cpu.c [hunk 5]
```diff
@@ -68,7 +67,9 @@ int hal_cpuDebugGuard(u32 enable, u32 slot)
 	else
 		cpu.dr5 &= ~mask;
 
-	__asm__ volatile ("movl %0, %%dr5" : : "r" (cpu.dr5));
+	__asm__ volatile(
+		"movl %0, %%dr5; "
+	:: "r" (cpu.dr5));
```
[the remark this answers is not in the document]
Comment (Maxez) [reply hal/ia32/cpu.c:72]: Using special formatting for asm code.

File: hal/ia32/cpu.c [hunk 6]
```diff
@@ -135,62 +135,36 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
...
+		"cli; "
+		"addl $12, %%esp; "
+
+		"call _interrupts_multilockClear; "
+
+		"jmp interrupts_popContext; "
+		"1: "
+	: "=a" (err)
+	: "g" (threads_schedule)
+	: "ecx", "edx", "esp", "cc", "memory");
```
[the remark this answers is not in the document]
Comment (Maxez) [reply hal/ia32/cpu.c:167]: Using special formatting for asm code.

File: hal/ia32/cpu.c [hunk 7]
```diff
@@ -250,7 +224,9 @@ void *_cpu_initCore(void)
 	cpu.tss[hal_cpuGetID()].esp0 = (u32)&cpu.stacks[hal_cpuGetID()][511];
 
 	/* Set task register */
-	__asm__ volatile ("ltr %%ax" : : "a" ((4 + cpu.ncpus) * 8));
+	__asm__ volatile(
+		"ltr %%ax; "
+	:: "a" ((4 + cpu.ncpus) * 8));
```
[the remark this answers is not in the document]
Comment (Maxez) [reply hal/ia32/cpu.c:229]: Using special formatting for asm code.

**Comment by Maxez** [comment]
I've added new commit with small changes (mostly formatting and changes proposed by @kemonats). I think the code looks good now.

**Review by Maxez — DISMISSED** [review]

File: hal/ia32/cpu.c [hunk 8]
```diff
@@ -135,62 +135,36 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
...
-
-		"jmp interrupts_popContext;"
-
-		"3:;"
-		"movl %%eax, %0"
-	: "=g" (err)
-	: "m" (spinlock), "m" (spinlock->lock), "r" (*scp), "g" (threads_schedule)
-	: "eax", "edx", "esp", "cc", "memory");
+	if (spinlock != NULL)
+		hal_spinlockClear(spinlock, scp);
```
Comment (agkaminski) [inline hal/ia32/cpu.c:139]: spinlock should be cleared only when reschedule is ready - see example here https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/6a2f9bc0bf3eba9829f0a4f271003fa141c0bad3/hal/armv7m/stm32/_init.S#L226 and here https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/6a2f9bc0bf3eba9829f0a4f271003fa141c0bad3/hal/armv7a/_interrupts.S#L129

So at least interrupts should be disabled before clearing spinlock.

But from what I see this implementation wasn't doing that anyway - spinlock clear on beginning, just like on this PR, so no regress here

File: syscalls.c [hunk 9]
```diff
@@ -802,8 +801,9 @@ int syscalls_signalPost(void *ustack)
 	err = threads_sigpost(proc, t, signal);
 
 	proc_put(proc);
-	threads_put(t);
-	hal_cpuReschedule(NULL, NULL);
+	if (t != NULL)
+		threads_put(t);
+
```
Comment (agkaminski) [inline syscalls.c:806]: `if` I get, but why remove reschedule?

Comment (Maxez) [reply syscalls.c:806]: Reschedule is already called inside `threads_sigpost()` - I think there is no need to call it again.

Comment (agkaminski) [reply syscalls.c:806]: Ok, makes sense

File: hal/ia32/cpu.c [hunk 10]
```diff
@@ -135,62 +135,37 @@ int hal_cpuReschedule(spinlock_t *spinlock, spinlock_ctx_t *scp)
...
- 		"pushl %%esp;"
-
-		"jmp interrupts_popContext;"
-
-		"3:;"
-		"movl %%eax, %0"
-	: "=g" (err)
-	: "m" (spinlock), "m" (spinlock->lock), "r" (*scp), "g" (threads_schedule)
-	: "eax", "edx", "esp", "cc", "memory");
+	hal_cpuDisableInterrupts();
```
Comment (agkaminski) [inline hal/ia32/cpu.c:138]: This won't work - spinlockClear will enable interrupts by restoring interrupt mask. That's why we need to do spinlockClear manually in reschedule - so we'll only clear lock without restoring interrupt mask

Comment (Maxez) [reply hal/ia32/cpu.c:138]: @agkaminski I see your point - `popf` in `hal_spinlockClear()` enables interrupts back, so I've restored manual spinlock clearing. Please take a look at the code again.

Sidenote: the original problem with `hal_cpuReschedule()` was passing "r" (*scp) in the input list. It caused page fault (NULL dereference) when calling `hal_cpuReschedule(NULL, NULL)`. The `*scp` dereference (red rectangle) got executed before `spinlock != NULL` check (`cmp $0x0, %eax` below the rectangle).
reschedule

File: hal/ia32/cpu.c [hunk 11]
```diff
@@ -250,7 +245,9 @@ void *_cpu_initCore(void)
 	cpu.tss[hal_cpuGetID()].esp0 = (u32)&cpu.stacks[hal_cpuGetID()][511];
 
 	/* Set task register */
-	__asm__ volatile ("ltr %%ax" : : "a" ((4 + cpu.ncpus) * 8));
+	__asm__ volatile(
+		"ltr %%ax; "
+	:: "a" ((4 + cpu.ncpus) * 8));
```
[the remark this answers is not in the document]
Comment (Maxez) [reply hal/ia32/cpu.c]: I agree. I've added the change.

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**